src/web/error.rs

```diff
@@ -0,0 +1,79 @@
+//! Error types for `web` modules.
+
+use std::fmt;
+
+use actix_web::http::header::ToStrError;
+use base64::DecodeError;
+use failure::{Backtrace, Context, Fail};
+use hawk::Error as ParseError;
+use hmac::crypto_mac::{InvalidKeyLength, MacError};
+use serde_json::Error as JsonError;
+
+use error::ApiError;
+
+/// An error occurred during HAWK authentication.
+#[derive(Debug)]
+pub struct HawkError {
+    inner: Context<HawkErrorKind>,
+}
+
+/// Causes of HAWK errors.
+#[derive(Debug, Fail)]
+pub enum HawkErrorKind {
+    #[fail(display = "{}", _0)]
+    Base64(#[cause] DecodeError),
+
+    #[fail(display = "expired payload")]
+    Expired,
+
+    #[fail(display = "{}", _0)]
+    Header(#[cause] ToStrError),
+
+    #[fail(display = "{}", _0)]
+    Hmac(MacError),
+
+    #[fail(display = "validation failed")]
+    InvalidHeader,
+
+    #[fail(display = "{}", _0)]
+    InvalidKeyLength(InvalidKeyLength),
+
+    #[fail(display = "{}", _0)]
+    Json(#[cause] JsonError),
+
+    #[fail(display = "missing header")]
+    MissingHeader,
+
+    #[fail(display = "missing id property")]
+    MissingId,
+
+    #[fail(display = "missing path")]
+    MissingPath,
+
+    #[fail(display = "missing \"Hawk \" prefix")]
+    MissingPrefix,
+
+    #[fail(display = "{}", _0)]
+    Parse(ParseError),
+
+    #[fail(display = "id property is too short")]
+    TruncatedId,
+}
+
+failure_boilerplate!(HawkError, HawkErrorKind);
+
+from_error!(DecodeError, ApiError, HawkErrorKind::Base64);
+from_error!(InvalidKeyLength, ApiError, HawkErrorKind::InvalidKeyLength);
+from_error!(JsonError, ApiError, HawkErrorKind::Json);
+from_error!(MacError, ApiError, HawkErrorKind::Hmac);
+from_error!(ToStrError, ApiError, HawkErrorKind::Header);
+from_error!(ParseError, ApiError, HawkErrorKind::Parse);
+
+impl From<HawkErrorKind> for ApiError {
+    fn from(kind: HawkErrorKind) -> Self {
+        let hawk_error: HawkError = Context::new(kind).into();
+        hawk_error.into()
+    }
+}
+
+unsafe impl Sync for HawkErrorKind {}
```

Comment:
Mentioned this on IRC: here's an alternative to the unsafe marker:

https://gist.github.com/pjenvey/c3b71ce0d684f07a6bf4b28f5e66b81b